base/compiler/optimize.jl

```diff
@@ -219,15 +209,15 @@ function stmt_effect_free(@nospecialize(stmt), @nospecialize(rt), src::Union{IRC
                         Any[argextype(args[i], src) for i = 2:length(args)])
             end
             contains_is(_PURE_BUILTINS, f) && return true
-            contains_is(_PURE_OR_ERROR_BUILTINS, f) || return false
+            contains_is(_EFFECT_FREE_BUILTINS, f) || return false
             rt === Bottom && return false
             return _builtin_nothrow(f, Any[argextype(args[i], src) for i = 2:length(args)], rt)
         elseif head === :new
             typ = argextype(args[1], src)
             # `Expr(:new)` of unknown type could raise arbitrary TypeError.
             typ, isexact = instanceof_tfunc(typ)
             isexact || return false
-            isconcretedispatch(typ) || return false
+            isconcretetype(typ) || return false
```

Comment:
Why do we want `iskindtype` to show up here as legal? It should not really matter (not formed by the frontend), but seems arbitrary

Reply:
@aviatesk ? I believe this is from your change.

Reply:
I think I made this change to avoid running `iskindtype` just because `iskindtype` won't appear within `:new` expression. But having it there might be safer. May I push changes.